// @metamask-private/delegation-deployments/dist/index.mjs
// src/index.ts
var CHAIN_ID = {
  mainnet: 1,
  optimism: 10,
  polygon: 137,
  base: 8453,
  arbitrum: 42161,
  linea: 59144,
  sepolia: 11155111,
  lineaGoerli: 59140,
  lineaSepolia: 59141,
  baseSepolia: 84532
};
var DELEGATOR_CONTRACTS = {
  "0.0.2": {
    [CHAIN_ID.sepolia]: {
      DelegationManager: "0xf57D1C4E5F460df42f8865046a7d87a62be0E495",
      EntryPoint: "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
      SimpleFactory: "0xD4536F3A97432eBF50F2B642CCaE1c05b673aBc8",
      // Implementations
      MultiSigDeleGator: "0xc63fc798A05b8862D1432323454A5151e2aa99Cd",
      OwnableDeleGator: "0x1C6DB47cC29667fA2ad3179fe48e84462F6fb47b",
      P256DeleGator: "0xeB1808907E9bFE622424b5AB8C41B528248624b8",
      // Caveat Enforcers
      AllowedMetadataEnforcer: "0x39ad96AFb5FDbC7452b2DA84D4d7F6c40e42d373",
      AllowedMethodsEnforcer: "0xf116fa8c365eaae30Ff70308E019c3B800FB1CB8",
      AllowedTargetsEnforcer: "0x6E4c779B4a6903F4B3C7239E7FCdD5435ce6F79c",
      BlockNumberEnforcer: "0x71Fb54d89fE50827749f45a76B01d79fA305e23F",
      DeployedEnforcer: "0xc5e290dC388e68b674171Ca097A559AFb26F8F41",
      ERC20AllowanceEnforcer: "0xF3b6054AE5ac9Bbe2e00339ACA46D3c6aD341509",
      EqualParameterEnforcer: "0xba808597432B781Fff84d709C4BA6cD66426b21b",
      LimitedCallsEnforcer: "0x4CE496Aed14427DB90328c97895998d1a0837f75",
      NonceEnforcer: "0x455e500487a79712F143622D5836db19ec347ecE",
      TimestampEnforcer: "0x58967Ce075DFE31e5b4aadFEa2854Ed9D539351c"
    }
  },
  "0.0.3": {
    [CHAIN_ID.sepolia]: {
      DelegationManager: "0x226cf2A6273fd6E544FB02880243a60e01F8ABBb",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0xba81a35e879010EaD82A503d637D30b520072EC8",
      // Implementations
      MultiSigDeleGatorImpl: "0x572fCF1f33126870cdFDC9935B161D6185Af865b",
      HybridDeleGatorImpl: "0x62576D59B701A7e488707Ed3e0980D2F37675Fd7",
      // Caveat Enforcers
      AllowedMetadataEnforcer: "0xD70eC675BE6F4e2F0240C6D3549e0C0B690b4B69",
      AllowedMethodsEnforcer: "0x50084B4c57764Cb5De78b721E5F634B6ce88C125",
      AllowedTargetsEnforcer: "0xDBc1C143FCe9aB810df9ECD231440D5367EA096b",
      BlockNumberEnforcer: "0xfaD6E73ADb22cCD096Ca2023c8Bdc15bf1C01f3F",
      DeployedEnforcer: "0x6c866cE0003F9A670304953b69DD1d9F8f8801A9",
      EqualParameterEnforcer: "0x50c9A75F75EDD7D69eFD4780845169A39f58F19c",
      ERC20AllowanceEnforcer: "0x50A010C86aCAc6B1C55e8a24532CdF3872bd811E",
      LimitedCallsEnforcer: "0xEa6cb0a20E94CeB692FFb89C455f1d4f9EAe0440",
      NonceEnforcer: "0x372DBD120571CDf1Ba86d63066C3b84bABB088FD",
      TimestampEnforcer: "0xe938aeD455f154c7B8fCA6e8A35a4329Ce018f87"
    },
    [CHAIN_ID.lineaGoerli]: {
      DelegationManager: "0x226cf2A6273fd6E544FB02880243a60e01F8ABBb",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0xba81a35e879010EaD82A503d637D30b520072EC8",
      // Implementations
      MultiSigDeleGatorImpl: "0x572fCF1f33126870cdFDC9935B161D6185Af865b",
      HybridDeleGatorImpl: "0x62576D59B701A7e488707Ed3e0980D2F37675Fd7",
      // Caveat Enforcers
      AllowedMetadataEnforcer: "0xD70eC675BE6F4e2F0240C6D3549e0C0B690b4B69",
      AllowedMethodsEnforcer: "0x50084B4c57764Cb5De78b721E5F634B6ce88C125",
      AllowedTargetsEnforcer: "0xDBc1C143FCe9aB810df9ECD231440D5367EA096b",
      BlockNumberEnforcer: "0xfaD6E73ADb22cCD096Ca2023c8Bdc15bf1C01f3F",
      DeployedEnforcer: "0x6c866cE0003F9A670304953b69DD1d9F8f8801A9",
      EqualParameterEnforcer: "0x50c9A75F75EDD7D69eFD4780845169A39f58F19c",
      ERC20AllowanceEnforcer: "0x50A010C86aCAc6B1C55e8a24532CdF3872bd811E",
      LimitedCallsEnforcer: "0xEa6cb0a20E94CeB692FFb89C455f1d4f9EAe0440",
      NonceEnforcer: "0x372DBD120571CDf1Ba86d63066C3b84bABB088FD",
      TimestampEnforcer: "0xe938aeD455f154c7B8fCA6e8A35a4329Ce018f87"
    },
    [CHAIN_ID.lineaSepolia]: {
      DelegationManager: "0xE6dB543DA2889362d35502fB2aA4f1Ad73D3d72e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0xb675aF15275D3235BADaeBAe380BB9A3C5F1442d",
      // Implementations
      MultiSigDeleGatorImpl: "0x3D56a1C9Ad8cCfeaa953B44D70eceA9301372418",
      HybridDeleGatorImpl: "0x3FADFb7Ff10d83761F2e36a1890AaD2f8E4F6EdF",
      // Caveat Enforcers
      AllowedTargetsEnforcer: "0x7072EE3d1993f67fe9ED2697978E8064cFA70ffa",
      AllowedMethodsEnforcer: "0xB3eF8eAEd53eBfd2B454E216670EDE39a081F9ff"
    }
  },
  "1.0.0": {
    // Mainnets
    [CHAIN_ID.optimism]: {
      DelegationManager: "0xbe4138886cb096bdc1b930f2f0ca7892aa234d78",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      // Implementations
      MultiSigDeleGatorImpl: "0x11f555af5844d85bfcf5d61d2a22866527eb585a",
      HybridDeleGatorImpl: "0xd6edd1256deccb2b06bdecef92dc16bcf26e531b",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x48db3835a873d64a4af2c09f014052407c003bd7",
      AllowedMethodsEnforcer: "0xfd731951bf1c52afccee3e6f14ab656475b76dd4",
      AllowedTargetsEnforcer: "0xbc8673c0afa52d86d991c06881e55b2966920564",
      BlockNumberEnforcer: "0xc15faffa0d879b9263c15a46ce31eacfa2e0e8ae",
      DeployedEnforcer: "0x5accb9559b56a6c1e3f90e342c85c42d93720d43",
      ERC20BalanceGteEnforcer: "0xb5d6b1ec6d868a3bae5b7f48178eaa2686a7a087",
      ERC20TransferAmountEnforcer: "0x92ac423b9c111962179a6242e1adb58d02c103be",
      IdEnforcer: "0x34152d9f3f8f74338d50703e780389e829b4abac",
      LimitedCallsEnforcer: "0x4b3adad4a328bee8ba17b86074d92fe7372180cd",
      NonceEnforcer: "0x2f32ff3fc3086d7f63f16fe8d0065390d460b40d",
      TimestampEnforcer: "0x78e05f779490c24bf3bfa135b4112e7003b321cd",
      ValueLteEnforcer: "0xfc20ede0a1132e839fbda9d7ed3904ff3c89540f"
    },
    [CHAIN_ID.polygon]: {
      DelegationManager: "0xbe4138886cb096bdc1b930f2f0ca7892aa234d78",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      // Implementations
      MultiSigDeleGatorImpl: "0x11f555af5844d85bfcf5d61d2a22866527eb585a",
      HybridDeleGatorImpl: "0xd6edd1256deccb2b06bdecef92dc16bcf26e531b",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x48db3835a873d64a4af2c09f014052407c003bd7",
      AllowedMethodsEnforcer: "0xfd731951bf1c52afccee3e6f14ab656475b76dd4",
      AllowedTargetsEnforcer: "0xbc8673c0afa52d86d991c06881e55b2966920564",
      BlockNumberEnforcer: "0xc15faffa0d879b9263c15a46ce31eacfa2e0e8ae",
      DeployedEnforcer: "0x5accb9559b56a6c1e3f90e342c85c42d93720d43",
      ERC20BalanceGteEnforcer: "0xb5d6b1ec6d868a3bae5b7f48178eaa2686a7a087",
      ERC20TransferAmountEnforcer: "0x92ac423b9c111962179a6242e1adb58d02c103be",
      IdEnforcer: "0x34152d9f3f8f74338d50703e780389e829b4abac",
      LimitedCallsEnforcer: "0x4b3adad4a328bee8ba17b86074d92fe7372180cd",
      NonceEnforcer: "0x2f32ff3fc3086d7f63f16fe8d0065390d460b40d",
      TimestampEnforcer: "0x78e05f779490c24bf3bfa135b4112e7003b321cd",
      ValueLteEnforcer: "0xfc20ede0a1132e839fbda9d7ed3904ff3c89540f"
    },
    [CHAIN_ID.base]: {
      DelegationManager: "0xbe4138886cb096bdc1b930f2f0ca7892aa234d78",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      // Implementations
      MultiSigDeleGatorImpl: "0x11f555af5844d85bfcf5d61d2a22866527eb585a",
      HybridDeleGatorImpl: "0xd6edd1256deccb2b06bdecef92dc16bcf26e531b",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x48db3835a873d64a4af2c09f014052407c003bd7",
      AllowedMethodsEnforcer: "0xfd731951bf1c52afccee3e6f14ab656475b76dd4",
      AllowedTargetsEnforcer: "0xbc8673c0afa52d86d991c06881e55b2966920564",
      BlockNumberEnforcer: "0xc15faffa0d879b9263c15a46ce31eacfa2e0e8ae",
      DeployedEnforcer: "0x5accb9559b56a6c1e3f90e342c85c42d93720d43",
      ERC20BalanceGteEnforcer: "0xb5d6b1ec6d868a3bae5b7f48178eaa2686a7a087",
      ERC20TransferAmountEnforcer: "0x92ac423b9c111962179a6242e1adb58d02c103be",
      IdEnforcer: "0x34152d9f3f8f74338d50703e780389e829b4abac",
      LimitedCallsEnforcer: "0x4b3adad4a328bee8ba17b86074d92fe7372180cd",
      NonceEnforcer: "0x2f32ff3fc3086d7f63f16fe8d0065390d460b40d",
      TimestampEnforcer: "0x78e05f779490c24bf3bfa135b4112e7003b321cd",
      ValueLteEnforcer: "0xfc20ede0a1132e839fbda9d7ed3904ff3c89540f"
    },
    [CHAIN_ID.arbitrum]: {
      DelegationManager: "0xbe4138886cb096bdc1b930f2f0ca7892aa234d78",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      // Implementations
      MultiSigDeleGatorImpl: "0x11f555af5844d85bfcf5d61d2a22866527eb585a",
      HybridDeleGatorImpl: "0xd6edd1256deccb2b06bdecef92dc16bcf26e531b",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x48db3835a873d64a4af2c09f014052407c003bd7",
      AllowedMethodsEnforcer: "0xfd731951bf1c52afccee3e6f14ab656475b76dd4",
      AllowedTargetsEnforcer: "0xbc8673c0afa52d86d991c06881e55b2966920564",
      BlockNumberEnforcer: "0xc15faffa0d879b9263c15a46ce31eacfa2e0e8ae",
      DeployedEnforcer: "0x5accb9559b56a6c1e3f90e342c85c42d93720d43",
      ERC20BalanceGteEnforcer: "0xb5d6b1ec6d868a3bae5b7f48178eaa2686a7a087",
      ERC20TransferAmountEnforcer: "0x92ac423b9c111962179a6242e1adb58d02c103be",
      IdEnforcer: "0x34152d9f3f8f74338d50703e780389e829b4abac",
      LimitedCallsEnforcer: "0x4b3adad4a328bee8ba17b86074d92fe7372180cd",
      NonceEnforcer: "0x2f32ff3fc3086d7f63f16fe8d0065390d460b40d",
      TimestampEnforcer: "0x78e05f779490c24bf3bfa135b4112e7003b321cd",
      ValueLteEnforcer: "0xfc20ede0a1132e839fbda9d7ed3904ff3c89540f"
    },
    [CHAIN_ID.linea]: {
      DelegationManager: "0xbe4138886cb096bdc1b930f2f0ca7892aa234d78",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      // Implementations
      MultiSigDeleGatorImpl: "0x11f555af5844d85bfcf5d61d2a22866527eb585a",
      HybridDeleGatorImpl: "0xd6edd1256deccb2b06bdecef92dc16bcf26e531b",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x48db3835a873d64a4af2c09f014052407c003bd7",
      AllowedMethodsEnforcer: "0xfd731951bf1c52afccee3e6f14ab656475b76dd4",
      AllowedTargetsEnforcer: "0xbc8673c0afa52d86d991c06881e55b2966920564",
      BlockNumberEnforcer: "0xc15faffa0d879b9263c15a46ce31eacfa2e0e8ae",
      DeployedEnforcer: "0x5accb9559b56a6c1e3f90e342c85c42d93720d43",
      ERC20BalanceGteEnforcer: "0xb5d6b1ec6d868a3bae5b7f48178eaa2686a7a087",
      ERC20TransferAmountEnforcer: "0x92ac423b9c111962179a6242e1adb58d02c103be",
      IdEnforcer: "0x34152d9f3f8f74338d50703e780389e829b4abac",
      LimitedCallsEnforcer: "0x4b3adad4a328bee8ba17b86074d92fe7372180cd",
      NonceEnforcer: "0x2f32ff3fc3086d7f63f16fe8d0065390d460b40d",
      TimestampEnforcer: "0x78e05f779490c24bf3bfa135b4112e7003b321cd",
      ValueLteEnforcer: "0xfc20ede0a1132e839fbda9d7ed3904ff3c89540f"
    },
    // Testnets
    [CHAIN_ID.sepolia]: {
      DelegationManager: "0xbe4138886cb096bdc1b930f2f0ca7892aa234d78",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0x11f555af5844d85bfcf5d61d2a22866527eb585a",
      // TODO: Replace with address deployed from known deployer.
      HybridDeleGatorImpl: "0x5989F5D13DF8fc818EdA65e417AED90459fD67F7",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x48db3835a873d64a4af2c09f014052407c003bd7",
      AllowedMethodsEnforcer: "0xfd731951bf1c52afccee3e6f14ab656475b76dd4",
      AllowedTargetsEnforcer: "0xbc8673c0afa52d86d991c06881e55b2966920564",
      BlockNumberEnforcer: "0xc15faffa0d879b9263c15a46ce31eacfa2e0e8ae",
      DeployedEnforcer: "0x5accb9559b56a6c1e3f90e342c85c42d93720d43",
      ERC20BalanceGteEnforcer: "0xb5d6b1ec6d868a3bae5b7f48178eaa2686a7a087",
      ERC20TransferAmountEnforcer: "0x92ac423b9c111962179a6242e1adb58d02c103be",
      IdEnforcer: "0x34152d9f3f8f74338d50703e780389e829b4abac",
      LimitedCallsEnforcer: "0x4b3adad4a328bee8ba17b86074d92fe7372180cd",
      NonceEnforcer: "0x2f32ff3fc3086d7f63f16fe8d0065390d460b40d",
      TimestampEnforcer: "0x78e05f779490c24bf3bfa135b4112e7003b321cd",
      ValueLteEnforcer: "0xfc20ede0a1132e839fbda9d7ed3904ff3c89540f"
    },
    [CHAIN_ID.lineaSepolia]: {
      DelegationManager: "0xbe4138886cb096bdc1b930f2f0ca7892aa234d78",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0x11f555af5844d85bfcf5d61d2a22866527eb585a",
      // TODO: Replace with address deployed from known deployer.
      HybridDeleGatorImpl: "0x5989F5D13DF8fc818EdA65e417AED90459fD67F7",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x48db3835a873d64a4af2c09f014052407c003bd7",
      AllowedMethodsEnforcer: "0xfd731951bf1c52afccee3e6f14ab656475b76dd4",
      AllowedTargetsEnforcer: "0xbc8673c0afa52d86d991c06881e55b2966920564",
      BlockNumberEnforcer: "0xc15faffa0d879b9263c15a46ce31eacfa2e0e8ae",
      DeployedEnforcer: "0x5accb9559b56a6c1e3f90e342c85c42d93720d43",
      ERC20BalanceGteEnforcer: "0xb5d6b1ec6d868a3bae5b7f48178eaa2686a7a087",
      ERC20TransferAmountEnforcer: "0x92ac423b9c111962179a6242e1adb58d02c103be",
      IdEnforcer: "0x34152d9f3f8f74338d50703e780389e829b4abac",
      LimitedCallsEnforcer: "0x4b3adad4a328bee8ba17b86074d92fe7372180cd",
      NonceEnforcer: "0x2f32ff3fc3086d7f63f16fe8d0065390d460b40d",
      TimestampEnforcer: "0x78e05f779490c24bf3bfa135b4112e7003b321cd",
      ValueLteEnforcer: "0xfc20ede0a1132e839fbda9d7ed3904ff3c89540f"
    }
  },
  "1.1.0": {
    // Mainnets
    [CHAIN_ID.arbitrum]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    },
    [CHAIN_ID.base]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    },
    [CHAIN_ID.linea]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    },
    [CHAIN_ID.optimism]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    },
    [CHAIN_ID.polygon]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    },
    // Testnets
    [CHAIN_ID.sepolia]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    },
    [CHAIN_ID.lineaSepolia]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    },
    [CHAIN_ID.baseSepolia]: {
      DelegationManager: "0x56D56e07e3d6Ee5a24e30203A37a0a460f42D7A3",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0xE8eA1DE8D6AfE400B7C8C1A81B7C29B7876b4d02",
      // Implementations
      MultiSigDeleGatorImpl: "0xd1f421EDbA5e3FA9efe3874827114b20C5BEC40C",
      HybridDeleGatorImpl: "0x941f3a016F8726d5643Ce62452d0D78492D42b42",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0xff71d60f3208469cBCE0859717B5198042DCB3F3",
      AllowedMethodsEnforcer: "0xe32C2561792e8446Abe73B9f557B881C13906186",
      AllowedTargetsEnforcer: "0x06aaE4c67EEA95277c46Bf79b1583d4a01772D22",
      BlockNumberEnforcer: "0x8E470D2Ae278457b42d2405E0B8Cd4BE21Ed9045",
      DeployedEnforcer: "0xf9088f013dBD9ebb7Cebd66fEB48253c6Ac5a820",
      ERC20BalanceGteEnforcer: "0xB7B6f32ec6343261D814e55Ed8C5925d91Cab861",
      ERC20TransferAmountEnforcer: "0x9A069b18032B31429A363AeCFb1B6A0564b44471",
      IdEnforcer: "0x91015c3b9D9523966eD2399885e5Df7A567f916c",
      LimitedCallsEnforcer: "0xe694bFfffEA3E85923b1210b37e6a0175e910863",
      NonceEnforcer: "0xE83BCFD8bBE672A96747e831050a91cf44F4F87A",
      TimestampEnforcer: "0x550FdD13eEBC1f22ea2a2480024BacBF0Ad7e5CE",
      ValueLteEnforcer: "0xBE32a6DB7471F63BB168C088c57Db01AfAe87967",
      NativeTokenTransferAmountEnforcer: "0x5eD3833d7B957A8DB8A461c3AF2d668Ec25382E0",
      NativeBalanceGteEnforcer: "0x376a98860E210DdEda3689fb39565592c563cB0A",
      ArgsEqualityCheckEnforcer: "0x7378dE585998d3E18Ce147867C335C25B3dB8Ee5",
      NativeTokenPaymentEnforcer: "0x87Fe18EbF99e42fcE8A03a25F1d20E119407f8e7",
      RedeemerEnforcer: "0x926672b130D1EF60A9d6b11D2048d121b30f40C1"
    }
  },
  "1.2.0": {
    // Mainnets
    [CHAIN_ID.arbitrum]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    },
    [CHAIN_ID.base]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    },
    [CHAIN_ID.linea]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    },
    [CHAIN_ID.optimism]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    },
    [CHAIN_ID.polygon]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    },
    // Testnets
    [CHAIN_ID.sepolia]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    },
    [CHAIN_ID.lineaSepolia]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    },
    [CHAIN_ID.baseSepolia]: {
      DelegationManager: "0x739309deED0Ae184E66a427ACa432aE1D91d022e",
      EntryPoint: "0x0000000071727De22E5E9d8BAf0edAc6f37da032",
      SimpleFactory: "0x6ff518884f21168c30c58CB21184D6AdBC18Ad90",
      // Implementations
      MultiSigDeleGatorImpl: "0xB4ab520FF1761f7b6dc221fEaCaf79367629Ed12",
      HybridDeleGatorImpl: "0xf4E57F579ad8169D0d4Da7AedF71AC3f83e8D2b4",
      // Caveat Enforcers
      AllowedCalldataEnforcer: "0x1021300501f6aDc446d4e506053F55a8a63cB1d7",
      AllowedMethodsEnforcer: "0x371f95c92Be3A916B824A2aE086Ed6db7A6193Fb",
      AllowedTargetsEnforcer: "0x91e043a13c61f9ddC02BDfe38dCA02A7F5b7Cc88",
      BlockNumberEnforcer: "0x955C7732562c6Dc4760dF749440f3ab28F46F608",
      DeployedEnforcer: "0xd2c8c04E2070c13CCB97FEAa25D1915676AAC191",
      ERC20BalanceGteEnforcer: "0x262A37B51798c44F1BEAd1A076703E4488887b78",
      ERC20TransferAmountEnforcer: "0x272E92835B12F014353E8754808C67682e9dddFA",
      IdEnforcer: "0xd6403989C2cc145102c2AE76E70D1317947ef587",
      LimitedCallsEnforcer: "0xA45dd3D90447640eB76085637132a74E18b310E3",
      NonceEnforcer: "0x1ba53a54eDa7021E08065C1C1943bCE91e0FceA3",
      TimestampEnforcer: "0xABc2591a40db08eD7045D91A29B3DBC33082DB54",
      ValueLteEnforcer: "0x9C458b17Cd6570e322Ee9a4180b309dAFD08e24C",
      NativeTokenTransferAmountEnforcer: "0xcfD1BD7922D123Caa194597BF7A0073899a284Df",
      NativeBalanceGteEnforcer: "0xDb5BAF405159f47Ab70DD424021ef114A450E101",
      ArgsEqualityCheckEnforcer: "0xACEC09a804020B307eFF00df9AAfb1Cf656DF9Cf",
      NativeTokenPaymentEnforcer: "0x6e21bABB3779bc694DC3A4DCeB35C1ecC1d9087b",
      RedeemerEnforcer: "0x596CC67C2Da64ED51E27B6d61f46e3F687E9182d",
      ERC721BalanceGteEnforcer: "0xd0960BC7324235DE4FAFe0f2eDCff64313220CC8",
      ERC721TransferEnforcer: "0x82621E65240f67D8F60a920F709127743A8D20A9",
      ERC1155BalanceGteEnforcer: "0x01a84C60B0B5c3EbB504fDa60a8236eB7e2D6655",
      OwnershipTransferEnforcer: "0x5f263261676d24Dd146545F22E485708900B2B83"
    }
  }
};
export {
  CHAIN_ID,
  DELEGATOR_CONTRACTS
};
